src/3d/mesh/qgsmesh3dmaterial_p.cpp

```diff
@@ -220,3 +311,56 @@ void QgsMesh3dMaterial::configure()
   mTechnique->addParameter( new Qt3DRender::QParameter( "isScalarMagnitude", ( mMagnitudeType == QgsMesh3dMaterial::ScalarDataSet ) ) );
 }
 
+void QgsMesh3dMaterial::configureArrows( QgsMeshLayer *layer )
+{
+  if ( !layer )
+    return;
+
+  QgsMeshDatasetIndex datasetIndex = layer->rendererSettings().activeVectorDataset();
+
+  mTechnique->addParameter( new Qt3DRender::QParameter( "arrowsSpacing", float( mSymbol.arrowsSpacing() ) ) ) ;
+  QColor arrowsColor = layer->rendererSettings().vectorSettings( datasetIndex.group() ).color();
+  mTechnique->addParameter( new Qt3DRender::QParameter( "arrowsColor", QVector4D( arrowsColor.redF(), arrowsColor.greenF(), arrowsColor.blueF(), 1.0f ) ) ) ;
+
+  QgsMeshDatasetGroupMetadata meta = layer->dataProvider()->datasetGroupMetadata( datasetIndex );
+  mTechnique->addParameter( new Qt3DRender::QParameter( "maxVectorLength", float( meta.maximum() ) ) ) ;
+
+  QVector<QgsVector> vectors;
+  QSize gridSize;
+  QgsPointXY minCorner;
+  Qt3DRender::QParameter *arrowsEnabledParameter = new Qt3DRender::QParameter( "arrowsEnabled", false );
+  if ( mMagnitudeType != MagnitudeType::ScalarDataSet || !mSymbol.arrowsEnabled() )
+    arrowsEnabledParameter->setValue( false );
+  else
+  {
+    arrowsEnabledParameter->setValue( true );
+    gridSize = QSize( mSymbol.maximumTextureSize(), mSymbol.maximumTextureSize() );
+    vectors = QgsMeshLayerUtils::griddedVectorValues(
+                layer,
+                datasetIndex,
+                mSymbol.arrowsSpacing(),
+                mSymbol.arrowsSpacing(),
+                gridSize,
+                minCorner );
+  }
+
+  mTechnique->addParameter( arrowsEnabledParameter )  ;
+
+  Qt3DRender::QTexture2D *arrowsGridTexture = new Qt3DRender::QTexture2D( this );
+  arrowsGridTexture->addTextureImage( new ArrowsGridTexture( vectors, gridSize ) );
+  arrowsGridTexture->setMinificationFilter( Qt3DRender::QTexture2D::Linear );
+  arrowsGridTexture->setMagnificationFilter( Qt3DRender::QTexture2D::Linear );
```

Comment:
shouldn't we use nearest neighbor approach here? I am wondering if this isn't going to create weird artifacts on the arrows if two neighboring cells have different angles?

Reply:
True